### ISSUE TITLE
docker image prune --all removes in-use image

### ISSUE THREAD

**Issue by githubsaturn** [moby/moby#36295]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

**Describe the results you received:**
Running `docker image prune --all --force` removes IN-USE images causing locally created images to get lost.

**Describe the results you expected:**
If a service is running via `locally-create-image:1`, running `docker image prune --all --force` SHOULD NOT remove `locally-create-image:1`

**Steps to reproduce the issue:**
1. `docker swarm init`     # simply create a swarm
2. `mkdir /test-bug`
3. `cd /test-bug`
4. `echo "FROM nginx:1.13.8" > Dockerfile`   # this is an example, it can be anything
5. `docker build -t myimage:1 . `
6. `docker service create --name my-service myimage:1`
7. wait and make sure that your container is running. Check using `docker ps -a`, you will see image is listed as `myimage:1`
8. `docker image prune --all --force`    # this is supposed to remove all "unused" images
9. `docker service update my-service --force`     # this can be an update of environmental variables or anything else that causes a container restart.
10. You will see that the service does not get restarted as there is no `myimage:1` image!

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:11:19 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:09:53 2017
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Output of `docker info`:**

```
Containers: 7
 Running: 4
 Paused: 0
 Stopped: 3
Images: 66
Server Version: 17.12.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 113
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: hl08z3yjk34msargicyuklvzy
 Is Manager: true
 ClusterID: k883g1werpi8dj207nr09kj3h
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 174.7.187.204
 Manager Addresses:
  174.7.187.204:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-112-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.682GiB
Name: REMOVED
ID: REMOVED
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: dockersaturn
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by dnephin** [comment]
The issue seems to be that some images are untagged by prune, even when they are used.

**Comment by dnephin** [comment]
This was modified in 1f7a9b1ab3d261de5be7d490e7e4f978f317242f  so cc @dmcgowan as well.

**Comment by tonistiigi** [comment]
@dnephin I think the logic in that function means that if an image has multiple references and one reference is deleted it just untags it, in any case, doesn't matter if there are containers or not.

**Comment by dnephin** [comment]
Yes, but that condition is a filter for when to check if the container exists. 

Why would it remove a tag that is used, just because there is more than one tag? Doesn't that sound broken?

**Comment by dnephin** [comment]
Apparently this very bizarre behaviour is even documented: 
https://github.com/moby/moby/blob/8f6a40a3f126ae1dc0a74dc612a3f7b14de4e7f6/daemon/image_delete.go#L35-L39

I'm not sure why, but apparently this is the expected behaviour.

**Comment by tonistiigi** [comment]
@dnephin I guess it so that `docker run busybox && docker tag busybox busybox2 && docker rmi busybox2` doesn't error

**Comment by dnephin** [comment]
Sure, but  `docker rmi busybox` should not untag `busybox`, however it does today. That's why I think the behaviour is wrong.

**Comment by tonistiigi** [comment]
A container isn't connected to a reference but the image itself by ID as references can always change during the lifetime of the container (I know there is a text field in config but it should be unused apart from container inspect view).

**Comment by dnephin** [comment]
I understand that, I don't think that changes my expectation for the behaviour. Also this issue is about services, which do use the reference when updated.

I would argue this is issue is a bug with ImageDelete, not specific to prune.

**Comment by tonistiigi** [comment]
> Also this issue is about services, which do use the reference when updated.

With the current code/api/ux, a container or `rmi` has no concept of a cluster.

> I would argue this is issue is a bug with ImageDelete, not specific to prune.

I think this was the desired behavior for `ImageDelete`. Is it the best solution or not is of course arguable. The base issue seems to be that someone thought all cases can be covered without the need for `untag` and that assumption doesn't seem to hold. But the bug, in this case, seems to be that prune uses the same codepath that `rmi` although it expects a different behavior than what is documented as the intended behavior for `rmi`.

**Comment by dnephin** [comment]
> the bug, in this case, seems to be that prune uses the same codepath that rmi although it expects a different behavior 

I disagree. `image prune` and `image rm` should absolutely use the same code path and have the exact same behaviour. There is no good reason for them to behave differently in some very subtle way. That is a terrible UX.

That leaves us with two options:
1. `ImageDelete` is wrong, but no one has really cared until now because it was only subtly wrong.
2. the current behaviour of `docker prune` is correct.

I personally think it is the first case.

**Comment by tonistiigi** [comment]
> image prune and image rm should absolutely use the same code path and have the exact same behaviour.

`rmi` is a combination of untag and delete image. This is not subtle but very intentional decision from the first 0.* releases to have an effect that references are same as images. You can't just skip the reference check in prune(that are different as no reference is explicit) and expect `rmi` to error when an image wasn't deleted because it will just untag then. Even if we fix this container ID vs image ID check (#36346) and this starts to work again it is almost accidental that the rmi code works here as most of it will be impossible to reach from prune. The bug in `ImageDelete` is that it should be called `ImageDeleteOrUntag` as it can return successfully without any image being deleted (as is clearly stated in the code comments and `rmi` docs).

**Comment by dnephin** [comment]
> rmi is a combination of untag and/or delete image. This is not subtle but very intentional decision

I agree, this is absolutely correct.

The problem is not that it untags images, the problem is what it untagged is not intuitive. When it was just the container API this was not a bit deal, but with the service API and `service update` it is more of an issue (which is why this github issue was opened). My reasoning for why this is not just a problem with prune is because the same problem exists with `image rm`:

```
$ docker image ls
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
myimage             1                   d3117424aaee        2 weeks ago         27.1MB
redis               alpine              d3117424aaee        2 weeks ago         27.1MB
$ docker service create myimage:1
...
lbeozsxxzluvrtkqvdqgumxho
$ docker container ls
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
50a94c08cf19        myimage:1           "docker-entrypoint..."   10 seconds ago      Up 9 seconds        6379/tcp            serene_volhard.1.t434m24w64enows81ed5034r3
$ docker image rm myimage:1
Untagged: myimage:1
$ docker service update --force lbeozsxxzluvrtkqvdqgumxho
overall progress: 0 out of 1 tasks
1/1: No such image: myimage:1
service update paused: update paused due to failure or early termination of task 
$
```

Exact same problem with `image rm`, the service is left in a broken state even though no `--force` was used on `rm`. This is exactly the same issue.

The problem is, as I mentioned above, that a reference that is "in use" is allowed to be removed. It's not really relevant the container is actually using an image ID because the image reference needs to exist for a user to re run  the `docker run` command, or to run ` service update` command. So from the perspective of a user, it is still in use.

`ImageDelete` should only allow you to untag if nothing is "using" that reference. The number of references that share the same image ID should be irrelevant.

**Comment by tonistiigi** [comment]
> Exact same problem with image rm, the service is left in a broken state even though no --force was used on rm. This is exactly the same issue.

I don't think this is the same issue. The issue you are describing is about different semantics between services and containers. A container is tied to an image, while service is tied to a reference that should point to a highly available external storage(with a fallback case in a worker if it fails). This fallback only works if a worker node has been manually set up in a specific way to answer a possible cluster request, while the node itself has no knowledge of being in a cluster. You are conveniently mixing cluster-manager and local-node commands in that example but in reality, we need to consider these commands running on different machines. To correctly solve this containers vs services problem we need a local cluster-aware storage that keeps track of all service images without the need for an external registry.

Prune is wrong in this case regardless if you want to change the deletion rules(that were designed before there were services). You don't need to use a service for prune to untag your image that is being used by a container.

**Comment by dnephin** [comment]
> I don't think this is the same issue. 

Did you see my example? It is identical.

> You don't need to use a service for prune to untag your image that is being used by a container.

Yes, both prune and image rm are broken for both containers and services, it's just more obviously wrong with services.

> You are conveniently mixing cluster-manager and local-node commands in that example but in reality, we need to consider these commands running on different machines. 

It's just as valid to run them on a single machine. It needs to work for both.

I think you're making this problem way more complicated than it needs to be. It's simple. A user issues a request to delete unused images. It removes an image that the user considers in-use, and that our CLI/API shows as in-use. That sounds like a bug to me. Either the the CLI/API shouldn't show the image tag, or the image delete is wrong.

**Comment by tonistiigi** [comment]
>  A user issues a request to delete unused images. It removes an image that the user considers in-use, and that our CLI/API shows as in-use. That sounds like a bug to me

Exactly. Prune is broken. It is broken because of a typo that compared image IDs to container IDs. Let's fix that please.

If you want to change the semantics of `rmi` that have been there for a long time and are documented and do have clear use-cases, please create another issue/proposal to discuss that.

**Comment by dnephin** [comment]
> It is broken because of a typo that compared image IDs to container IDs. Let's fix that please.

Keeping a map of image IDs was never the right solution.  prune is fixed, now the only issue is with image remove. `image delete` and `image prune` must behave the same way. We don't need a separate issue.

**Comment by dmcgowan** [comment]
> `image delete` and `image prune` must behave the same way

Why? They are very different functionality, one the user is requesting to remove something specific, the second is the daemon deciding what might be eligible for removal.

We shouldn't break the `image delete` behavior to fix a bug in prune. We have been fairly consistent in not breaking this sort of behavior because it tends to have more unexpected consequences for users than the original behavior. Honestly, it is confusing either way so we just need to make it clear in documentation. I would rather avoid changing this behavior now as it will have an impact on any future changes involving referencing/tagging.

**Comment by dnephin** [comment]
`prune` should just be an automated version of `image rm`. That should be the only difference. Otherwise it's unnecessary confusion for the user.

In the case of `image rm` the engine still needs to figure out if it's eligible for removal as well. It's even documented here: 

https://github.com/moby/moby/blob/8f6a40a3f126ae1dc0a74dc612a3f7b14de4e7f6/daemon/image_delete.go#L47-L54

> We shouldn't break the image delete behavior to fix a bug in prune

We aren't. We're fixing a bug in `image rm`:

```
$ docker pull redis:alpine
...
$ docker tag redis:alpine myimage:1
$ docker run -d myimage:1
552ae8f1134235f9c7d41ec220c9978cb51fd387d4a2d2cd4215b108950ee7c1
$ docker container ls --format '{{.Image}}'
myimage:1
$ docker image rm myimage:1
Untagged: myimage:1
$ docker image rm redis:alpine
Error response from daemon: conflict: unable to remove repository reference "redis:alpine" (must force) - container 20c7d341f7f5 is using its referenced image d3117424aaee
$ docker container ls --format '{{.Image}}'
d3117424aaee
```

I don't know how you can say this isn't broken. It's fine to untag a reference that IS being used and referenced by a container, but it's not ok to remove a completely unreferenced tag?

If we have to keep a tag around, we should keep the tag that's actually being used.

**Comment by tonistiigi** [comment]
> What is the use case?

for example:
- tagging image with `myimage-backup` without removing the container and freeing up the reference to be used by the new container with an option to rollback to the old image
- clearing up junk images that are duplicates of used ones

> prune should just be an automated version of image rm. That should be the only difference. Otherwise it's unnecessary confusion for the user.

I explained the differences 3 comments above. Prune is about removing unused images, `rmi` is not only about removing images but also about untagging a specific reference that user has picked out.

Those untagging rules are running before determining conflicts and documented just 10 lines before the conflicts documentation.

> I don't know how you can say this isn't broken.

It isn't broken because this is what it was designed to do and how it is documented. That design idea/compromise was that container is not related to a reference but an actual image. Changing a reference after creating a container has no effect to the container after that, it is only a pointer for finding an image on creation time. `rmi` isn't only way to untag an image, you can do the same thing with `tag` and `pull`. Your example is much easier to understand than these cases. User explicitly asks to remove or untag the image. It is successfully untagged. It should not be a surprise at all that it can't be used in a completely new `docker run`. Container list shows a link to parent image so all the tools using that link still work. The ref doesn't disappear from the container list because the reference was deleted but because the list code knows that references are mutable and always checks if whatever string user typed when starting a container would still be valid as a reference to that image. Inspect always shows image ID, and inspecting images shows all current refs as equal.

In case you have better ideas feel free to make proposals but finding a documented behavior that has weird edge cases does not make it a bug. `rmi` behaves like documented, `prune` does not behave like documented.

**Comment by dnephin** [comment]
> tagging image with myimage-backup without removing the container and freeing up the reference to be used by the new container with an option to rollback to the old image

This is still supported by my proposed behaviour because you can always `build` or `image tag` over an existing tag. You don't even need to delete the original tag. 

> clearing up junk images that are duplicates of used ones

Should still be supported as long as no container is using them. If a container is using them they aren't junk, right? Or if you really want to you can always `-f`.

> untagging myrepo:latest when you know that it is not up-to-date anymore. so that when new containers are created they would point to/pull a new image

This is the only use case that isn't directly supported. However a more straighforward way of doing this would be to `docker pull myrepo:latest`. Why just `rmi` the thing instead of updating it directly?

>It isn't broken because this is what it was designed to do and how it is documented

That's fair. Broken is the wrong word for it. I think it's unexpected behaviour that needs to be fixed, so you're correct that I should open a new issue for this. I have opened #36435.

I still maintain that `image prune` and `image rm` should behave the same way (with the exception of one being automated). Not because `image rm` should act like `prune`, but because consistent and expected behaviour is an important quality for an API. So I consider this issue (#36295) as a wont-fix, as it will be addressed by #36435.

**Comment by vilic** [comment]
We have images tagged with both `:latest` and `:[version]`, and we use `:[version]` in our compose file. After pruning the images, restart the server/docker, the services then just refuse to work (without re-deploying).

This is certainly wrong.

**Comment by thaJeztah** [comment]
copying from https://github.com/docker/cli/issues/2247#issuecomment-572479531

I agree that for `docker image prune -a` this is definitely confusing;  
If an image is in use, and tagged under multiple names, then running `docker image prune` will remove tags for the image that is in used but will keep 1 tag. Not sure if the tags that are removed are "random" or _explicitly_ include the tag that's actually in use, but in the example below, the tag that's actually _used_ is removed, and tags that are _unused_ are removed.

If possible, I think that that behaviour (at least for `docker image prune`) should be changed.

### example

Pull the `busybox:latest` image, and tag the image under some other names:

```bash
docker image pull busybox:latest
docker image tag busybox:latest  thajeztah/busybox:latest
docker image tag busybox:latest  thajeztah/busybox2:latest
```

there's now three tags for the same image

```bash
docker image ls
REPOSITORY           TAG                 IMAGE ID            CREATED             SIZE
busybox              latest              6d5fcfe5ff17        13 days ago         1.22MB
thajeztah/busybox2   latest              6d5fcfe5ff17        13 days ago         1.22MB
thajeztah/busybox    latest              6d5fcfe5ff17        13 days ago         1.22MB
```

Now,  run a container, using the `busybox:latest` tag

```bash
docker run -dit --name mycontainer busybox:latest
```

Prune images with the `-a/--all` option set

```bash
docker image prune -a
WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: busybox:latest
untagged: busybox@sha256:6915be4043561d64e0ab0f8f098dc2ac48e077fe23f488ac24b665166898115a
untagged: thajeztah/busybox2:latest

Total reclaimed space: 0B
```

Notice that in this case, it removed the `thajeztah/busybox2:latest` and `busybox:latest` tags (even though the container was using that), but keeps one tag so that the image still is referenced.

Removing the container, and running `docker image prune -a` again will remove the remaining tag for the image, and will remove the image itself:

```bash
docker rm -f mycontainer

WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: thajeztah/busybox:latest
deleted: sha256:6d5fcfe5ff170471fcc3c8b47631d6d71202a1fd44cf3c147e50c8de21cf0648
deleted: sha256:195be5f8be1df6709dafbba7ce48f2eee785ab7775b88e0c115d8205407265c5

Total reclaimed space: 1.22MB
```

To make the behavior less surprising, I think it should (for `docker image prune`) remove the _unused_ tag(s), and keep the tags that are in use.

**Comment by robinstrom** [comment]
Maybe some helpful input: 
This was working as expected in Docker API version 1.13 but currently not on version 1.39. 

Tried to look into the source code but I'm far too inexperienced in Golang to be able to help more unfortunately. Hope it is solved soon since it's a great future when working :)

**Comment by eduarddrenth** [comment]
Think I ran into this issue as well, I build and deploy images like so:

```bash
for dir in onfw fhwb fnwb nfwb jurwb
do
  cd $dir/dockerapp
  export APPNAME=$dir
  docker build --build-arg EXISTADDONSVERSION=2.5-SNAPSHOT --secret id=adminpw,src=adminpw -t ${APPNAME}:${VERSION} .
  docker stack rm $APPNAME
  sleep 15
  docker stack deploy --compose-file docker-compose.yml $APPNAME
  cd ../..
done
```
`image ls` will now show 5 images
Regularly I run `docker system prune -a`
The 4 images last created are removed, athough in use.

**Comment by sanchayanghosh** [comment]
Update as of now:

I tried running a dockers service create in this way, as I will show below - 

pull an image  - example node

`
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker pull node:latest
latest: Pulling from library/node
1cfaf5c6f756: Pull complete 
c4099a935a96: Pull complete 
f6e2960d8365: Pull complete 
dffd4e638592: Pull complete 
2ce46ad6f244: Pull complete 
5fda1f69c996: Pull complete 
b084d682aa98: Pull complete 
3ee73259434e: Pull complete 
a2aaca92485c: Pull complete 
Digest: sha256:d75b1663c035a5e9ed48ad54041aff3067d6a66104d32d07901314e573393d2a
Status: Downloaded newer image for node:latest
docker.io/library/node:latest
root@6304ba57e6c6:/go/src/github.com/docker/docker# 

`
At this point I will add a tag - 

`
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker tag node:latest node:test-latest
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker images
REPOSITORY   TAG           IMAGE ID       CREATED      SIZE
node         latest        c66552d59c4b   3 days ago   907MB
node         test-latest   c66552d59c4b   3 days ago   907MB
root@6304ba57e6c6:/go/src/github.com/docker/docker# 


`
I will now do a swarm init

`
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker swarm init
Swarm initialized: current node (87zgxm7bmb7atzyxi7y1dia70) is now a manager.

To add a worker to this swarm, run the following command:

    docker swarm join --token SWMTKN-1-1pow8rv58dbox6pyhksuqmhma75huoan91mxz458igy3ivngtw-1nkizwg7by9hkv5tcxellw1t2 172.17.0.2:2377

To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.

root@6304ba57e6c6:/go/src/github.com/docker/docker# 


`

I will now create a service with the image I pulled and tagged

`
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker service create --name test-swarm-part node:test-latest
image node:test-latest could not be accessed on a registry to record
its digest. Each node will access node:test-latest independently,
possibly leading to different nodes running different
versions of the image.

j80gua52w9pnhapiqdobybcrq
overall progress: 0 out of 1 tasks 
1/1: preparing [=================================>                 ] 
verify: Detected task failure 



`

I will now interrupt this so it runs in the background and will initiate a docker prune.

`
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker service ps j80gua52w9pnhapiqdobybcrq
ID             NAME                    IMAGE              NODE           DESIRED STATE   CURRENT STATE                     ERROR     PORTS
pwj51qbirehi   test-swarm-part.1       node:test-latest   6304ba57e6c6   Running         Starting less than a second ago             
xc6mo2lfio2m    \_ test-swarm-part.1   node:test-latest   6304ba57e6c6   Shutdown        Complete 7 seconds ago                      
pfh1qldq2mlm    \_ test-swarm-part.1   node:test-latest   6304ba57e6c6   Shutdown        Complete 15 seconds ago                     
t8rzanguefjx    \_ test-swarm-part.1   node:test-latest   6304ba57e6c6   Shutdown        Complete 23 seconds ago                     
yxk30l0lguf2    \_ test-swarm-part.1   node:test-latest   6304ba57e6c6   Shutdown        Complete 30 seconds ago                     
root@6304ba57e6c6:/go/src/github.com/docker/docker# 



`

`
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker system prune -f -a
Deleted Containers:
c315569f40c86cf86ffcaef57335a9819384bf63b1b939638c7cbd03ab5de400
814f60104f319d53e31ca3f5c0e95f568c8b79e5683b45b083d9a46df8e8eed6
b0f9a929eb3f21baeaef8f23db1c8dece358d07043858d66a64d71c4b1bf3531
48debf46ff0265e3346152a07756be9ffaac22af7466fae80a5c2867f54971e4

Deleted Images:
untagged: node:latest
untagged: node:test-latest
untagged: node@sha256:d75b1663c035a5e9ed48ad54041aff3067d6a66104d32d07901314e573393d2a
deleted: sha256:c66552d59c4bb478baea52cbcb218071bd7e4741d22ffa9ab541e805f03cdb1d
deleted: sha256:8a1dcbd7489c84bc09212471e28dc200c98497fc3374a0c6871198dfa850d462
deleted: sha256:3cfd4e2396fc95630c673fdcbce48ee376050cd623a125e3392b69f75b167e20
deleted: sha256:b7cddaf35158ec0a157c5c4bef5d5d3331a5c94a543a77eb0a5a8d202c40eb32
deleted: sha256:15c4781b83e9ed45b43e126beeeb211b76e08162ded1fdbc834acef583d0fc2f
deleted: sha256:ce2b1446aa6bbacae8efa1dcb7fca77ea1b09039112bd9829bf02eca2fe75bd8
deleted: sha256:a84b4e945affc390a49fe16b1e47251671dfc628c111de3a4c5b2e0b09d2649a
deleted: sha256:4bc4490d37ab1d9664581bcbb9c198cd5c4e7eaf0c5b1607c5ec44a11eaf922f
deleted: sha256:21b5c6b4b2751d9e70a183749330b4b9e11687a0cb4995f33ee5812846128294
deleted: sha256:c2ddc1bc2645ab5d982c60434d8bbc6aecee1bd4e8eee0df7fd08c96df2d58bb

Total reclaimed space: 906.9MB
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker ps -a
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker service ps -a
unknown shorthand flag: 'a' in -a
See 'docker service ps --help'.
root@6304ba57e6c6:/go/src/github.com/docker/docker# docker images
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE
root@6304ba57e6c6:/go/src/github.com/docker/docker# 



`

Here, as you may observe, both images will deleted even though they were neither dangling nor unused. They were in use by the service container.

I will investigate this issue further and issue a bug fix. Meanwhile, if this is documented, do let me know. However, this is not august with the documented behaviour on docker docuemntation.

**Comment by thejasn** [comment]
Hey @thaJeztah 

After further investigation looks like the issue is over here,
https://github.com/moby/moby/blob/2f74fa543b2f9ed6a3fd7c96afe3faa57b45f7e7/daemon/images/image_delete.go#L83-L97

The `isSingleReference()` is iterating through the references in the order of creation without checking if any container exists, since `isSingleReference()` returns a `false` we proceed to remove the references. 

The above-mentioned behavior doesn't happen when using `docker image prune [IMAGE_ID]` since it relies on a different set of checks before deleting and moreover relies on `imageDeleteHelper()`.
https://github.com/moby/moby/blob/2f74fa543b2f9ed6a3fd7c96afe3faa57b45f7e7/daemon/images/image_delete.go#L157-L164

Also while looking into this, found a small discrepancy in behavior between the 2 flow, 
- `docker image rm [IMAGE_ID]` gives the below error if a container attached with the same image is running.  (doesn't allow to be forced)
```
root@a46e2988ae4f:/go/src/github.com/docker/docker# docker ps -a
DEBU[2021-08-22T21:31:20.412062382Z] Calling GET /_ping                           
DEBU[2021-08-22T21:31:20.412332961Z] Calling GET /v1.30/containers/json?all=1     
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
19d35ebceb80        busybox:latest      "sh"                12 minutes ago      Up 1 second                             mycontainer

----------------------------------------------

Error response from daemon: conflict: unable to delete 42b97d3c2ae9 (cannot be forced) - image is being used by running container 19d35ebceb80
```
- `docker image rm [IMAGE_NAME]` gives the below error in the same scenario, (allows to be forcibly removed successfully)
```
Error response from daemon: conflict: unable to remove repository reference "busybox" (must force) - container 19d35ebceb80 is using its referenced image 42b97d3c2ae9
```

### Fix
- To fix the issue at hand I'm thinking of having the check mentioned here, out into the if expression so as to give priority for running containers. 
https://github.com/moby/moby/blob/master/daemon/images/image_delete.go#L89
- Since the second discrepancy isn't documented anywhere (couldn't find one at least), assuming it's an issue I think we should re-use the same checks built across the two variations (by ID ref and repo ref). 

Please let me know if this approach is valid, will push the bug fix after that.

**Comment by eikaramba** [comment]
would be great if this issue can be addressed finally

**Comment by davidavdav** [comment]
I would also appreciate it the issue is addressed

**Comment by MasterMind2k** [comment]
I have similar problem.

I've prepared a work-around (it is a bit noisy, but it works):
```
/usr/bin/docker image ls -q | /usr/bin/xargs /usr/bin/docker image rm
```

In my case, problem happens when single image has two tags. I am using git commit as an docker image tag, it could happen that two commits have same image. Prune will (randomly?) untag one of tags.

**Comment by borbonchristian** [comment]
any update on this topic?

**Comment by DKroot** [comment]
Our team had our main Production service brought down because of this behavior. I have played with this and want to clarify what's going on.
We're on Docker CE version 28.0.4, build b8034c0, RHEL 9.

As explained in [Docker docs](https://docs.docker.com/reference/cli/docker/image/rm/), when an image is tagged with multiple tags `docker image remove {tag}` (a.k.a. `docker rmi`) would untag a specified tag but leave the image available in Docker. This is reasonable.
However, `docker system prune—-all` cleanup removes all tags except the **last one in alphabetical order** when an image is tagged with multiple tags.

Here is a test case:
```
docker run --detach -it --rm busybox:latest

docker ps -a
# Output:
#CONTAINER ID   IMAGE                 COMMAND             CREATED         STATUS         PORTS     NAMES
#926d4479930e   busybox:latest        "sh"                8 seconds ago   Up 8 seconds             frosty_ramanujan

docker tag busybox:latest busybox:foo
docker tag busybox:latest busybox:qux

docker images
# Output:
#REPOSITORY   TAG            IMAGE ID       CREATED        SIZE
#busybox      foo            ff7a7936e930   7 months ago   4.28MB
#busybox      latest         ff7a7936e930   7 months ago   4.28MB
#busybox      qux            ff7a7936e930   7 months ago   4.28MB

docker system prune -a

docker ps -a
# Output:
#CONTAINER ID   IMAGE                 COMMAND             CREATED         STATUS         PORTS     NAMES
#926d4479930e   ff7a7936e930          "sh"                2 minutes ago   Up 2 minutes             frosty_ramanujan

docker images
# Output:
#REPOSITORY   TAG            IMAGE ID       CREATED        SIZE
#busybox      qux            ff7a7936e930   7 months ago   4.28MB
```

This behavior has nothing to do with what image tag is listed in `docker ps`, and it is potentially dangerous: it might remove a tag you rely on despite what the docs claim ("unused images"). The containers continue functioning as long as the loaded image is available by its ID. Still, when scripts operate with containers, for example, doing periodic restarts, they typically rely on a specific image tag. And that tag might be gone. A practical advice is to avoid using multiple tags altogether until this defect is fixed. When an image, for example, needs to be pushed to a different private Distribution Registry, tag it correspondingly but remove the original tag.

I definitely consider this a defect because this behavior in  `docker system prune—-all` makes no sense, and it might bring your Production down if you use it on occasion.

**Comment by DKroot** [comment]
@thaJeztah It went like this:
1. `docker service create --name app-prod app:1.2.3`. The image happened to have two tags.
2. Nightly maintenance script: `docker system prune -a`
3. Periodic restart: `docker service update --force --image app:1.2.3 app-prod`. Kaboom. No such tag. There were other complications: the private registry was down, but this was the crux of the issue.
